### PR TITLE
Implement first data scanner logic

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: src/main/java/life/qbic/data/processing/AppConfig.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      @Value("${scanner.directory}") String scannerDirectory,
+      @Value("${scanner.interval}") int interval) {
+    return new ScannerConfiguration(scannerDirectory, interval);
+  }
+
+  @Bean
+  RegistrationWorkersConfig registrationWorkersConfig(
+      @Value("${registration.threads}") int amountOfWorkers,
+      @Value("${registration.working.dir}") String workingDirectory,
+      @Value("${registration.target.dir}") String targetDirectory) {
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/AppConfig.java:38]: There is only one target dir - is this correct as their are multiple steps in the process?

Comment (wow-such-code) [reply src/main/java/life/qbic/data/processing/AppConfig.java:38]: nevermind :)

File: src/main/java/life/qbic/data/processing/Provenance.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  public static final String FILE_NAME = "provenance.json";
+
+  /**
+   * The path from where the dataset has been picked up originally.
+   */
+  @JsonProperty("origin")
+  public String originPath;
+
+  @JsonProperty("user")
+  public String userPath;
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/Provenance.java:37]: what path is this?

File: src/main/java/life/qbic/data/processing/Provenance.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * The current location identifier of the dataset
+   * <p>
+   * Can be a file path when operating on file system level
+   */
+  @JsonProperty("history")
+  public List<String> history;
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/Provenance.java:45]: why is it a list, then?

File: src/main/java/life/qbic/data/processing/config/RegistrationWorkersConfig.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+public class RegistrationWorkersConfig {
+
+  private final int amountOfWorkers;
+
+  private final Path workingDirectory;
+
+  private final Path targetDirectory;
+
+  public RegistrationWorkersConfig(int amountOfWorkers, String workingDirectory, String targetDirectory) {
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/config/RegistrationWorkersConfig.java:14]: NIT: this is defined as "threads" in other configs

File: src/main/java/life/qbic/data/processing/AppConfig.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        measurementIdPattern);
+  }
+
+  @Bean
+  EvaluationConfiguration evaluationConfiguration(EvaluationWorkersConfig evaluationWorkersConfig,
+      GlobalConfig globalConfig) {
+    return new EvaluationConfiguration(evaluationWorkersConfig.workingDirectory().toString(),
+        evaluationWorkersConfig.targetDirectory().toString(),
+        evaluationWorkersConfig.measurementIdPattern().toString(), globalConfig);
+  }
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/AppConfig.java:65]: these seem to be doing mostly the same - what is the idea behind that?

File: src/main/java/life/qbic/data/processing/scanner/Scanner.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      ConcurrentRegistrationQueue registrationQueue, GlobalConfig globalConfig) {
+    this.setName("Scanner-Thread");
+    Objects.requireNonNull(scannerConfiguration, "scannerConfiguration must not be null");
+    scannerPath = Path.of(scannerConfiguration.scannerDirectory());
+    if (!scannerPath.toFile().exists()) {
+      throw new RuntimeException("Could not find scanner directory: " + scannerPath);
+    }
+    this.scanInterval = scannerConfiguration.scanInterval();
+    this.registrationQueue = Objects.requireNonNull(registrationQueue,
+        "eventQueue must not be null");
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/scanner/Scanner.java:56]:
```suggestion
        "registrationQueue must not be null");
```

File: src/main/java/life/qbic/data/processing/scanner/Scanner.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+  private List<RegistrationRequest> detectDataForRegistration() {
+    return userProcessDirectories.parallelStream()
+        .map(Path::toFile)
+        .map(file -> createRequests(file.listFiles(), file.toPath())).flatMap(
+            Collection::stream).toList();
+  }
+
+  private boolean isNotErrorFolder(File folder) {
+    return !folder.getName().equals(usersErrorDirectory.getFileName());
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/scanner/Scanner.java:101]: doesn't this check the error folder itself and not its content? this seems like a very specific fail case (user moves the error folder?)

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -1 +1,55 @@
...
+# Data processing app properties
+#================================
+
+#------------------------
+# Global settings
+#------------------------
+# Directory name that will be used for the manual intervention directory
+# Created in the users' home folders
+# e.g. /home/<user1>/error
+users.error.directory.name=error
```
Comment (wow-such-code) [inline src/main/resources/application.properties:11]: NIT: I would place this below the scanner.directory

File: src/main/java/life/qbic/data/processing/ConcurrentRegistrationQueue.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package life.qbic.data.processing;
+
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import life.qbic.data.processing.registration.RegistrationRequest;
+
+/**
+ * <b><class short description - 1 Line!></b>
```
Comment (wow-such-code) [inline src/main/java/life/qbic/data/processing/ConcurrentRegistrationQueue.java:8]: missing javadoc

File: src/main/java/life/qbic/data/processing/Provenance.java (hunk 3 above)
Comment (wow-such-code) [reply src/main/java/life/qbic/data/processing/Provenance.java:45]: (documentation does not fit imho)

File: src/main/java/life/qbic/data/processing/Provenance.java (hunk 2 above)
Comment (sven1103) [reply src/main/java/life/qbic/data/processing/Provenance.java:37]: Yes exactly, the user's working directory, if you want. I try to find a more suitable semantic

File: src/main/resources/application.properties (hunk 8 above)
Comment (sven1103) [reply src/main/resources/application.properties:11]: it is a global setting and not specific for the scanner part, so I favor it to be separated :D

**Review by wow-such-code — APPROVED** [review]
nice

**Review by wow-such-code — APPROVED** [review]

**Pull request merged**